src/main/java/com/es/phoneshop/DAO/CustomProductDao.java

```diff
@@ -57,21 +63,84 @@ public List<Product> findProducts() {
         } finally {
             readLock.unlock();
         }
+    }
+
+    public List<Product> findProductsByQuery(String query) {
+        readLock.lock();
+        try {
+            Comparator<Product> comparator = setQueryComparator(query);
+            return getProductList(query, comparator);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    private List<Product> getProductList(String query, Comparator<Product> comparator) {
+        return products.stream()
+                .filter(product -> StringUtils.isEmpty(query) ||
+                        StringChecker.calculateStringSimilarity(product.getDescription(), query) > 0)
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getPrice().compareTo(BigDecimal.valueOf(0)) > 0)
+                .filter(product -> product.getStock() > 0)
+                .sorted(comparator)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<Product> sortProductsByFieldAndQuery(SortField sortField, SortOrder sortOrder, String query) {
+        readLock.lock();
+        try {
+            Comparator<Product> comparator = setOrderComparator(sortOrder,
+                    setFieldComparator(sortField, query,
+                            setQueryComparator(query)));
+            //i saved query here because in filter page we must sort by query(relevance), not by description it is not logical
+            return getProductList(query, comparator);
+        } finally {
+            readLock.unlock();
+        }
 
     }
 
+    private Comparator<Product> setQueryComparator(String query) {
+        if (StringUtils.isEmpty(query)) {
+            return Comparator.naturalOrder();
```

Comment:
if query is empty we can simply return not sorted products from findProducts()

Reply:
but here the comparator returns, not the list of sorted/unsorted products